### ISSUE TITLE
Feature: AWS STS support

### ISSUE THREAD

**Issue by matthdsm** [igvteam/igv.js#1084]
Hi,

Are there any plans to support AWS STS endpoints? This would be quite useful next to the oauth support for people who have data in secure buckets and want to view data using temporary signed URLS.

If so, any way we could help drive this?

Thanks
M

**Comment by northwestwitch** [comment]
Hi @matthdsm I made a recent pull request in my repo that does something like that: https://github.com/Clinical-Genomics/scout/pull/1841 , I'm not using security token but AWS access_key and secret_access_key

**Comment by matthdsm** [comment]
Hi @northwestwitch, 
Lookin' good! 
That's about the same as we have in mind, but written in JS.
We're looking into using the [AWS.WebIdentityCredentials](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/WebIdentityCredentials.html) to access a private minio cluster cfr the [example on the minio github](https://github.com/minio/minio/blob/master/docs/sts/web-identity.md)

It'd be awesome if this was natively supported by igv.js. I can see a whole lot of usecases where this feature would be handy.

M

**Comment by jrobinso** [comment]
@matthdsm @northwestwitch could you be more precise on the workflow and what you would need for igv.js to do, as opposed to what its containing application might do?  You know you can set arbitrary headers including oAuth tokens now, correct?    I'm just not sure what you mean by "natively support".

**Comment by matthdsm** [comment]
Hi Jim!
Thanks for the reply.
The flow for STS is pretty much the same as for "regular" oAuth.
It pretty much consists of:
```
redirect to oAuth -> access_code
access_code -> access_token
access_token - fetch credentials from aws -> temporary S3 credentials
```
Temp credentials can then be used to get data from a secure bucket.

Do you thinks that's something that can be handled within igv.js? Or is this something the surrounding application should handle?
Is there an option to set a function as `url` for a track?

You mention arbitrary headers. Is this documented?

Thanks 
M

**Comment by jrobinso** [comment]
Generally oAuth interactions are handled by the application,  and the token passed to igv.js.   igv.js is just a component on the page along with other components.  igv-webapp OTOH is an application and handles oAuth.

RE STS, that does not look like "regular" oAuth, at least that's not a google flow.   There is an extra step "fetch credentials from aws...".  With Google, and MS, once you have an access token you can access the resources with an "Authorization: bearer" header.   What are these "temporary S3 credentials" and how exactly are they used?  An example might help.

The idea to use use a function as a url is interesting, or more likely a Promise.  Then you can do async work and resolve the promise when its done.   This is an option for the oAuth property now, it could also be made to work for URLs fairly easily.   Would that allow you to implement your flow? 

RE headers, yes its  documented

**Comment by jrobinso** [comment]
@matthdsm While not explicitly about STS,  I am experimenting with allowing a function or promise for url properties.   See the branch "function_urls", and example below. 

Note the oauthToken already supports functions and promises, in addition to a string value.  I am inferring from your request that the oAuth token is not enough in this Amazon flow, you need to use it to fetch some other credentials.  If you can tell me how these credentials are used that would help.   I should educate myself but have very little time (e.g. are they used in request headers, or are they signed urls, or something else).

```
            tracks: [
                {
                    name: "CTCF - string url",
                    type: "wig",
                    format: "bigwig",
                    url: "https://www.encodeproject.org/files/ENCFF563PAW/@@download/ENCFF563PAW.bigWig"
                },
                {
                    name: "CTCF - function url ",
                    type: "wig",
                    format: "bigwig",
                    url: function () {
                        return "https://www.encodeproject.org/files/ENCFF563PAW/@@download/ENCFF563PAW.bigWig"
                    }
                },
                {
                    name: "CTCF - promise url",
                    type: "wig",
                    format: "bigwig",
                    url: Promise.resolve("https://www.encodeproject.org/files/ENCFF563PAW/@@download/ENCFF563PAW.bigWig")
                }
            ]
```

**Comment by matthdsm** [comment]
Hi Jim,

To clarify what I'm aiming at here. 
Most AWS services support an STS endpoint, which allows user without an amazon account to perform actions on a buckets.
Instead of using an AWS account or the access/secret key for said bucket, the user can then log in by "assuming" a web identity through an openID connect provider (e.g. google, facebook, homegrown implementation). 
The STS endpoint returns temporary credentials which can in turn be used to fetch objects.

So long story short, I'm looking for something that allows users to "log in" into an AWS repo using their oAuth credentials, which returns a temporary URL to be used in igv.js.

Hope this clarified some of your questions.

All that being said, the url function implementation looks great. If you think the full STS flow is out of scope for igv.js, this will already help me a great deal!


Thanks a lot!
Cheers,
M

**Comment by jrobinso** [comment]
The full STS is out of scope for now simply because I'm overwhelmed with other things.   It could possibly be added in the future.   The functional URL approach might not be enough unless you can included the temporary credentials as url parameters.   If they need to be added as headers we will need a bit more.   Let me know how it works, you'll need to pull from the branch.   I am likely to release this anyway as I think it could be used for lots of applications.

**Comment by matthdsm** [comment]
The approach I’m thinking is to use the temporary credentials to generate a signed url, which can then be used as a track. Thats where the track url as a function could come in. 
The rest can be handled by the surrounding application as far as I’m concerned. 

Thanks for bearing with me on this!
Cheers
M

**Comment by jrobinso** [comment]
Yes that should work.    Use a Promise if this is asynchronous, and resolve it when you get the signed URL,  or reject if the authentication fails.     Also, as currently coded the url or Promise will be invoked on every request,  this is on purpose in case credentials have a time limit and need refreshed.   So on your side you'll want to cache the signed url and just return or resolve it in future calls.   I thought about invoking the function/promise once, but this would be less flexible.

**Comment by matthdsm** [comment]
Sound great!
Thanks a lot for the help.

Cheers
M

**Comment by brainstorm** [comment]
@matthdsm @northwestwitch Instead of tapping directly into STS, may I suggest looking at Cognito instead? It should handle your usecase and beyond, securely, please take a peek at @UMCCR's work with the desktop version over here:

https://umccr.org/blog/igv-amazon/

Since the above is already in upstream IGV desktop, I think it would make sense to implement a similar mechanism on IGV.js too?

Cheers!

**Comment by jrobinso** [comment]
Thanks @brainstorm !   There is an important difference between IGV desktop and igv.js,  IGV desktop is a complete application.    igv.js is a genomics viewing component intended to be embedded in a web application.   The enclosing web application is generally responsible for user management, including authentication.   What I try to do with igv.js is provide enough flexibility to support any authentication workflow.  

What would be really helpful to the community generally is an example application using a Cognito workflow with an embedded igv.js.   igv.js might have enough flexibility to support this already, we really wouldn't know for sure until someone tries.

**Comment by matthdsm** [comment]
Cognito might be something that would work, but the issue here, for me at least, is that we're not using amazon in any way.

We're using OIDC from [hydra](https://www.ory.sh/hydra/) to authenticate against a [minio](https://github.com/minio/minio) cluster on our own servers, cutting out the middle man. Hence the STS solution, as this is something minio does support.

Cheers
M

**Comment by matthdsm** [comment]
this was fixed by using the new url-as-a-promise functionality, using a function to generate a signed url.

Thanks for all the help!
Cheers
M
